### PR TITLE
Support nested_tensor.bmm

### PR DESCRIPTION
**Pull request by YifanShenSZ**

### CODE REVIEW

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 1]
```diff
@@ -713,5 +713,59 @@ Tensor softmax_nested(const Tensor& input, const int64_t dim, const bool half_to
...
+  auto self_ptr = get_nested_tensor_impl(self),
+      mat2_ptr = get_nested_tensor_impl(mat2);
+  int64_t ntensors = self_ptr->size(0),
+      ntensors2 = mat2_ptr->size(0);
+  TORCH_CHECK(ntensors == ntensors2,
+      "Expected size for the 1st dimension of batch2 tensor to be: ", ntensors,
+      " but got: ", ntensors2, ".");
+  const Tensor& self_sizemat = self_ptr->get_nested_size_tensor(),
+      & mat2_sizemat = mat2_ptr->get_nested_size_tensor();
+  TORCH_CHECK(ntensors > 0 && self_sizemat.size(1) == 2, "batch1 must be a 3D tensor");
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:726]: You can use ```self.dim() == 3``` and ```math2.dim() == 3```. I'd do this check at the very beginning.

File: test/test_nestedtensor.py [hunk 2]
```diff
@@ -578,6 +578,36 @@ def test_softmax(self, device, dtype):
...
         self.assertRaises(IndexError, lambda: torch.nn.functional.softmax(nt1, 1))
 
+    @dtypes(torch.float, torch.float16, torch.double)
+    @torch.inference_mode()
+    def test_bmm(self, device, dtype):
+        # error case: not 3D tensors
+        nt0 = torch.nested_tensor([])
+        nt1 = torch.nested_tensor([torch.randn(2), torch.randn(3)])
+        nt2 = torch.nested_tensor([torch.randn((2, 4)), torch.randn((3, 4))])
+        self.assertRaises(RuntimeError, lambda: nt0.bmm(nt0))
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:588]: Bonus points for `assertRaisesRegex` so we can make sure we get the error message (and thus the error) that we expect. This prevents bugs more strongly, since `RuntimeError` is such a common Error.

Comment (cpuhrsch) [reply test/test_nestedtensor.py:588]: You can also orient yourself on the error messages that torch.bmm throws for regular torch.Tensors

Comment (YifanShenSZ) [reply test/test_nestedtensor.py]: Rewritten the tests with `assertRaisesRegex`. 3D check error message is consistent.

The size check, however, I had to tweak a bit:
* regular torch.Tensor.bmm check the first 2 dimensions of batch2
* nested tensor has a well defined 1st dimension, but maybe not the 2nd dimension

See the `assertRaisesRegex` codes for the error messages I came up with

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 3]
```diff
@@ -713,5 +713,58 @@ Tensor softmax_nested(const Tensor& input, const int64_t dim, const bool half_to
...
+      mat2_shapes = NestedTensor_get_shapes(mat2_ptr);
+  const Tensor& self_buffer = self_ptr->get_buffer(),
+      & mat2_buffer = mat2_ptr->get_buffer();
+  // determine output size
+  const Tensor& self_sizemat = self_ptr->get_nested_size_tensor();
+  Tensor out_sizemat = self_sizemat.new_empty(self_sizemat.sizes());
+  int64_t* out_sizemat_ptr = out_sizemat.data_ptr<int64_t>();
+  std::vector<int64_t> out_offsets(ntensors + 1);
+  std::vector<IntArrayRef> out_shapes(ntensors);
+  out_offsets[0] = 0;
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:738]: I don't think this needs to be a std_vector. You're only using the `i-1`th value once to calculate the `i-th` value and then only use that last value to determine the size of `out_buffer`. You could just create a `int64_t out_numel` into which you accumulate the sizes of the output.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 4]
```diff
@@ -713,5 +713,58 @@ Tensor softmax_nested(const Tensor& input, const int64_t dim, const bool half_to
...
+  for (int64_t i = 0; i < ntensors; i++) {
+    const IntArrayRef& self_shape = self_shapes[i],
+        & mat2_shape = mat2_shapes[i];
+    const int64_t& self_size0 = self_shape[0], & self_size1 = self_shape[1],
+        & mat2_size0 = mat2_shape[0], & mat2_size1 = mat2_shape[1];
+    TORCH_CHECK(self_size1 == mat2_size0,
+        "Nested matrices cannot be multiplied (",
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:745]: We can be more precise about which nested matrix by printing the index `i` here.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 5]
```diff
@@ -713,5 +713,58 @@ Tensor softmax_nested(const Tensor& input, const int64_t dim, const bool half_to
...
+        mat2_size0, "x", mat2_size1, ")");
+    out_sizemat_ptr[0] = self_size0;
+    out_sizemat_ptr[1] = mat2_size1;
+    out_shapes[i] = IntArrayRef(out_sizemat_ptr, out_sizemat_ptr + 2);
+    out_sizemat_ptr += 2;
+    out_offsets[i + 1] = out_offsets[i] + self_size0 * mat2_size1;
+  }
+  Tensor out_buffer = self_buffer.new_empty(out_offsets.back());
+  // call tensor mm
+  // TODO: for cpu, maybe use `parallel_for`
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:756]: For common cases it'll be much more efficient, until we have a specialized kernel, to pad out the NestedTensors, call bmm on the padded Tensors and then remove the padding again. This will be a necessary follow-up task.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 3 above)
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:738]: `out_offsets` will be used again in line 761 to determine which segment of `out_buffer` catches the `i`th matmul result

**Review by cpuhrsch — APPROVED** [review]
Looks good, just see last comment about error message for incompatible nested sizes (matrices cannot be multiplied).

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 5 above)
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp]: marked in "TODO"

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -578,6 +578,48 @@ def test_softmax(self, device, dtype):
...
+        # error case: underlying matrices cannot be multiplied
+        nt0 = torch.nested_tensor([torch.randn((2, 4)), torch.randn((3, 4))])
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"0-th nested matrices in batch cannot be multiplied \(2x4 and 2x4\)",
+            lambda: nt0.bmm(nt0)
+        )
+        # normal nested tensor
+        nt0 = torch.nested_tensor([torch.randn((2, 4)), torch.randn((3, 4))])
+        nt1 = torch.nested_tensor([torch.randn((4, 6)), torch.randn((4, 5))])
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:618]: testing nit: use slightly different sizes so there's no way that nt0[0] and nt1[1] can be multiplied together, for example

**Review by jbschlosser — APPROVED** [review]
Looks great!
